src/felzenszwalb.jl

```diff
@@ -0,0 +1,184 @@
+"""
+```
+index_map, num_segments = felzenszwalb_graph(edges, num_vertices, k)
+```
+
+Efficient Graph-Based Image Segmentation.
+"""
+function felzenszwalb_graph(edges::Array{SVector{3, Float64}}, num_vertices::Int, k::Real)
+
+    num_edges = length(edges)
+    num_sets = num_vertices
+    rank = zeros(num_sets)
+    set_size = ones(num_sets)
+    parent = Array(range(1, num_sets))
+    threshold = fill(k, num_vertices)
+    
+    function find_root(x)
+        x = Int(x)
+        if parent[x]!=x
+            parent[x] = find_root(parent[x])
+        end
+        return parent[x]
+    end
+
+    function merge_trees(x, y)
+        x = Int(x)
+        y = Int(y)
+        num_sets -= 1
+        if rank[x]<rank[y]
+            parent[x] = y
+            set_size[y] += set_size[x]
+            return y
+        elseif rank[x]>rank[y]
+            parent[y] = x
+            set_size[x] += set_size[y]
+            return x
+        else
+            parent[y] = x
+            set_size[x] += set_size[y]
+            rank[x] += 1
+            return x
+        end
+    end
+
+    sort(edges, lt = (x,y)->(x[3]<y[3]))
+
+    for i in range(1,num_edges)
+        edge = edges[i]
+        a, b, w = edge
+        a = find_root(a)
+        b = find_root(b)
+        if a!=b
+            if w <= min(threshold[a], threshold[b])
+                merged_root = merge_trees(a, b)
+                threshold[merged_root] = w + k/set_size[merged_root]
+            end
+        end
+    end
+
+    segments = OrderedSet()
+    for i in 1:num_vertices
+        push!(segments, find_root(i))
+    end
+
+    segments2index = Dict{Int, Int}()
+    for i in 1:num_sets
+        segments2index[segments[i]]=i 
+    end
+
+    index_map = Array{Int64}(num_vertices)
+    for i in 1:num_vertices
+        index_map[i] = segments2index[find_root(i)]
+    end
+
+    return index_map, num_sets
+end
+
+function felzenszwalb{T<:Gray}(img::AbstractArray{T, 2}, k::Real; sigma=0.8)
```

Comment:
![s](https://user-images.githubusercontent.com/17504447/27601871-6bcf31a2-5b8e-11e7-8906-2a32f82730e2.jpg)
Without filtering, segmentation is noisy.
![s2](https://user-images.githubusercontent.com/17504447/27601872-6bd07da0-5b8e-11e7-957a-601c3159de4f.jpg)
I was uneasy adding the filtering, but added it only because the author said that gaussian filter with σ = 0.8 always gives good results. I looked at some open-sourced implementations and found that they post process the segmentation result to remove small components. After that the initial filtering shouldn't be necessary.